### PR TITLE
PIM-5902: clean and filter provider

### PR DESCRIPTION
**Pull request by juliensnz**
This PR clean small stuff on the peb. I added a filter provider to be able to customize the look of the filters (for us and integrators).

### CODE REVIEW

File: src/Pim/Bundle/EnrichBundle/Resources/config/form_extensions/filters.yml [hunk 1]
```diff
@@ -1,13 +1,13 @@
 extensions:
-    pim-filter-text:
+    akeneo-text-filter:
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Resources/config/form_extensions/filters.yml]: I don't understand those akeneo/pim renamings

I don't see how these things could be used outside the PIM as the attributes are currently a PIM notion

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterProviderInterface.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+namespace Pim\Bundle\EnrichBundle\Provider\Filter;
+
+/**
+ * Filter provider interface
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterProviderInterface.php]: explain what is does please, what's the purpose to provide filters? what for? where?

Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterProviderInterface.php]: and by the way I'm really not a big fan of the `filter` naming
I have the feeling it mixes too many things, but maybe I'm wrong

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/BaseFilterProvider.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+use Pim\Component\Catalog\Model\AttributeInterface;
+
+/**
+ * Filter provider for attributes
+ *
+ * @author    Julien Sanchez <julien@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class BaseFilterProvider implements FilterProviderInterface
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/BaseFilterProvider.php]: don't like Base
to me it means you have to override this class

File: src/Pim/Bundle/EnrichBundle/Normalizer/AttributeNormalizer.php [hunk 4]
```diff
@@ -57,6 +64,7 @@ public function normalize($attribute, $format = null, array $context = [])
             'wysiwyg_enabled'       => $attribute->isWysiwygEnabled(),
             'empty_value'           => $this->emptyValueProvider->getEmptyValue($attribute),
             'field_type'            => $this->fieldProvider->getField($attribute),
+            'filter_type'           => $this->filterProvider->getFilters($attribute),
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Normalizer/AttributeNormalizer.php]: `filter_types` maybe?

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/BaseFilterProvider.php (hunk 3 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/BaseFilterProvider.php]: any suggestion?

File: src/Pim/Bundle/EnrichBundle/Normalizer/AttributeNormalizer.php (hunk 4 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Normalizer/AttributeNormalizer.php]: Yeah, good idea

File: src/Pim/Bundle/EnrichBundle/Resources/config/form_extensions/filters.yml (hunk 1 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/config/form_extensions/filters.yml]: it's not linked to an attribute. And not related to a product either. We named them like that on the pef so I did this to be consistent. Tell me if you still want me to rename them

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterProviderInterface.php (hunk 2 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterProviderInterface.php]: any suggestion to replace filter?

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/BaseFilterProvider.php (hunk 3 above)
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/BaseFilterProvider.php]: `AttributeFilterProvider` ?

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterChainedProvider.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
+<?php
+
+namespace Pim\Bundle\EnrichBundle\Provider\Filter;
+
+/**
+ * Filter chained provider
+ *
+ * @author    Julien Sanchez <julien@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterChainedProvider.php]: 2k16

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterProviderInterface.php [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * @author    Julien Sanchez <julien@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterProviderInterface.php]: 2k16

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterProviderInterface.php (hunk 2 above)
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterProviderInterface.php]: @jjanvier You're right. Those filters are only for the UI. They provide the right "**UI module name for a given attribute filter**" for the front to display the right component. 

I'd like to have the "UI" notion somewhere in those classes (name or namespace)

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterChainedProvider.php [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class FilterChainedProvider implements FilterProviderInterface
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterChainedProvider.php]: Could be nice to tell in the name that this is to display UI filter components

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/filter/attribute/date.js [hunk 8]
```diff
@@ -134,19 +125,17 @@ define([
...
-                FetcherRegistry
-                    .getFetcher('attribute')
-                    .fetch(this.getCode())
+                FetcherRegistry.getFetcher('attribute').fetch(this.getCode())
             ).then(function (templateContext, attribute) {
                 return _.extend({}, templateContext, {
                     label: i18n.getLabel(attribute.labels, UserContext.get('uiLocale'), attribute.code)
                 });
-            });
+            }.bind(this));
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/filter/attribute/date.js]: not needed

File: src/Pim/Bundle/ReferenceDataBundle/Enrich/Provider/Filter/ReferenceDataFilterProvider.php [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+namespace Pim\Bundle\ReferenceDataBundle\Enrich\Provider\Filter;
+
+use Pim\Bundle\EnrichBundle\Provider\Filter\FilterProviderInterface;
+use Pim\Component\Catalog\Model\AttributeInterface;
+
+/**
+ * Filter provider for reference data
+ *
+ * @author    Julien Sanchez <julien@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
```
Comment (grena) [inline src/Pim/Bundle/ReferenceDataBundle/Enrich/Provider/Filter/ReferenceDataFilterProvider.php]: 2016

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterChainedProvider.php (hunk 7 above)
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterChainedProvider.php]: ChainedFilterProvider is better imo

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterProviderInterface.php (hunk 6 above)
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterProviderInterface.php]: 2016

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterChainedProvider.php (hunk 5 above)
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterChainedProvider.php]: 2016

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/BaseFilterProvider.php [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+namespace Pim\Bundle\EnrichBundle\Provider\Filter;
+
+use Pim\Component\Catalog\AttributeTypes;
+use Pim\Component\Catalog\Model\AttributeInterface;
+
+/**
+ * Filter provider for attributes
+ *
+ * @author    Julien Sanchez <julien@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/BaseFilterProvider.php]: 2016

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterProviderInterface.php [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface FilterProviderInterface
+{
+    /**
+     * Get the Filter for the given element
+     *
+     * @param mixed $element
+     *
+     * @throws RuntimeException If no filter is found for the given element
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterProviderInterface.php]: What about a domain exception?

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterChainedProvider.php [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+        return false;
+    }
+
+    /**
+     * Add a provider
+     *
+     * @param FilterProviderInterface $provider
+     */
+    public function addProvider(FilterProviderInterface $provider)
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterChainedProvider.php]: I don't understand when it's called :/

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/export/product/edit/content/data.js [hunk 13]
```diff
@@ -142,16 +140,14 @@ define(
...
-                    configProvider.getFilters('product-export-builder')
-                ).then(function (attribute, config) {
-                    return {
-                        field: attribute.code,
-                        view: config[attribute.type].view,
-                        isRemovable: true
-                    };
-                });
+                return fetcherRegistry.getFetcher('attribute').fetch(fieldCode)
+                    .then(function (attribute, config) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/export/product/edit/content/data.js]: `config` is not used anymore :fire: :dancer:

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/filter/attribute/identifier.js [hunk 14]
```diff
@@ -59,18 +50,11 @@ define([
...
-                .getFetcher('attribute')
-                .fetch(this.getField())
-                .then(function (attribute) {
-                    return {
-                        label: i18n.getLabel(
-                            attribute.labels,
-                            UserContext.get('catalogLocale'),
-                            this.getField()
-                        ),
+            return BaseFilter.prototype.getTemplateContext.apply(this, arguments)
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/filter/attribute/identifier.js]: I think this override can be entirely dropped

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/filter/attribute/metric.js [hunk 15]
```diff
@@ -95,7 +86,6 @@ define([
                 FetcherRegistry.getFetcher('measure').fetchAll()
             ).then(function (templateContext, attribute, measures) {
                 return _.extend({}, templateContext, {
-                    label: i18n.getLabel(attribute.labels, UserContext.get('uiLocale'), attribute.code),
                     units: measures[attribute.metric_family]
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/filter/attribute/metric.js]: You have the attribute in the context now, no need to fetch it anymore.
Maybe other cases like this in other filters, can you check?

File: upgrades/schema/Version_1_6_20160726103445_update_product_export_parameters.php [hunk 16]
```diff
@@ -53,7 +53,23 @@ public function up(Schema $schema)
...
+                            'field' => 'categories.code',
+                            'operator' => 'IN CHILDREN',
+                            'value' => [$channel->getCategory()->getCode()]
+                        ],
+                        [
+                            'field' => 'completeness',
+                            'operator' => '>=',
+                            'value' => 100
+                        ]
+                    ],
```
Comment (phaseinducer) [inline upgrades/schema/Version_1_6_20160726103445_update_product_export_parameters.php]: Nice catch!

**Comment by grena** [comment]
GTM once fixed and things renamed :dancer:

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/filter/attribute/metric.js (hunk 15 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/filter/attribute/metric.js]: good catch!

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/filter/attribute/identifier.js (hunk 14 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/filter/attribute/identifier.js]: how?

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterChainedProvider.php (hunk 12 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterChainedProvider.php]: here: https://github.com/akeneo/pim-community-dev/blob/master/src/Pim/Bundle/EnrichBundle/DependencyInjection/Compiler/RegisterGenericProvidersPass.php#L70 :)

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterProviderInterface.php (hunk 11 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterProviderInterface.php]: removed, it was not the case :)

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterChainedProvider.php (hunk 7 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterChainedProvider.php]: we named all our chained provider like that

File: src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterProviderInterface.php (hunk 2 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Provider/Filter/FilterProviderInterface.php]: It's in the enrich bundle so it's for the UI. I don't think we should add more

**Comment by phaseinducer** [comment]
GTM

**Pull request merged**